// arlas-web-components/esm2022/lib/components/results/model/item.mjs
/*
 * Licensed to Gisaïa under one or more contributor
 * license agreements. See the NOTICE.txt file distributed with
 * this work for additional information regarding copyright
 * ownership. Gisaïa licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
;
export class Item {
    constructor(columns, itemData) {
        /**
         * @description More data organized in groups.
         */
        this.itemDetailedData = new Array();
        /**
         * @description Whether to display the detailed data.
         */
        this.isDetailToggled = false;
        /**
         * @description Whether the item is checked.
         */
        this.isChecked = false;
        /**
         * @description Whether the item state is indeterminated.
         */
        this.isindeterminated = false;
        /**
         * @description Whether to highlight the item.
         */
        this.ishighLight = undefined;
        this.columns = columns;
        this.itemData = itemData;
    }
}
