// arlas-map/esm2022/lib/map-import/map-import.component.mjs
/*
 * Licensed to Gisaïa under one or more contributor
 * license agreements. See the NOTICE.txt file distributed with
 * this work for additional information regarding copyright
 * ownership. Gisaïa licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { Component, ElementRef, Inject, Input, Output, ViewChild } from '@angular/core';
import { MAT_DIALOG_DATA, MatDialog, MatDialogRef } from '@angular/material/dialog';
import * as toGeoJSON from '@tmcw/togeojson';
import centroid from '@turf/centroid';
import JSZip from 'jszip';
import { polygon } from '@turf/helpers';
import { Subject } from 'rxjs';
import * as shp_ from 'shpjs/dist/shp';
import * as extent_ from 'turf-extent';
import { parse } from 'wellknown';
import { valid } from 'geojson-validation';
import { marker } from '@colsen1991/ngx-translate-extract-marker';
import * as gpsi_ from 'geojson-polygon-self-intersections';
import { ArlasMapComponent } from '../arlas-map.component';
import { ArlasMapFrameworkService } from '../arlas-map-framework.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/material/dialog";
import * as i2 from "@angular/common";
import * as i3 from "@angular/forms";
import * as i4 from "@angular/material/button";
import * as i5 from "@angular/material/checkbox";
import * as i6 from "@angular/material/progress-spinner";
import * as i7 from "@angular/material/radio";
import * as i8 from "@ngx-translate/core";
import * as i9 from "../arlas-map-framework.service";
const gpsi = gpsi_.default;
const shp = shp_.default;
const extent = extent_.default;
export class MapImportDialogComponent {
    constructor(dialogRef, data) {
        this.dialogRef = dialogRef;
        this.data = data;
        this.displayError = false;
        this.isRunning = false;
        this.fitResult = false;
        this.wktContent = '';
        this.SHP = marker('shp');
        this.KML = marker('kml');
        this.WKT = marker('wkt');
        this.GEOJSON = marker('geojson');
        this.file = new Subject();
        this.importRun = new Subject();
        this.fitResult = data.defaultFitResult ?? false;
        this.allowedImportType = this.data.allowedImportType.filter(t => [this.SHP, this.KML, this.WKT, this.GEOJSON].includes(t));
        if (this.allowedImportType.indexOf(this.SHP) > -1) {
            this.importType = this.SHP;
        }
        else if (this.allowedImportType.indexOf(this.KML) > -1) {
            this.importType = this.KML;
        }
        else if (this.allowedImportType.indexOf(this.WKT) > -1) {
            this.importType = this.WKT;
        }
        else {
            this.importType = this.GEOJSON;
        }
        this.changeType();
    }
    onFileChange(files) {
        this.file.next(files.item(0));
        this.currentFile = files.item(0);
        this.displayError = false;
    }
    import() {
        this.importRun.next({ type: this.importType, fitResult: this.fitResult, wktContent: this.wktContent });
    }
    onTextChange() { }
    changeType() {
        if (this.importType === this.SHP) {
            this.allowedFileExtension = '.zip';
        }
        else if (this.importType === this.KML) {
            this.allowedFileExtension = '.kml,.kmz';
        }
        else if (this.importType === this.GEOJSON) {
            this.allowedFileExtension = '.json,.geojson';
        }
    }
    onCancel() {
        this.dialogRef.close();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: MapImportDialogComponent, deps: [{ token: i1.MatDialogRef }, { token: MAT_DIALOG_DATA }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "18.2.13", type: MapImportDialogComponent, selector: "arlas-map-import-dialog", outputs: { file: "file", importRun: "importRun" }, viewQueries: [{ propertyName: "fileInput", first: true, predicate: ["fileInput"], descendants: true }], ngImport: i0, template: "<h1 mat-dialog-title class=\"mapgl-import-title\">{{ 'Import file' | translate}}</h1>\n<div mat-dialog-content class=\"mapgl-import-body\">\n  <mat-radio-group class=\"mapgl-import-radio-group\" [(ngModel)]=\"importType\" (change)=\"changeType()\">\n    <mat-radio-button *ngFor=\"let type of allowedImportType\" [value]=\"type\" class=\"mapgl-import-radio-button\">\n      {{ type | translate }}\n    </mat-radio-button>\n  </mat-radio-group>\n\n  <input *ngIf=\"importType !== WKT\" type=\"file\" class=\"mapgl-import-upload\" #fileInput accept=\"{{allowedFileExtension}}\" (change)=\"onFileChange($event.target.files)\">\n  <textarea *ngIf=\"importType === WKT\" class=\"mapgl-import-text-area\" [(ngModel)]=\"wktContent\" placeholder=\"{{'Paste your WKT' | translate}}\" (change)=\"onTextChange()\"></textarea>\n\n  <div class=\"mapgl-import-zoom\">\n    <mat-checkbox [(ngModel)]=\"fitResult\">{{ 'Zoom to result' | translate}}</mat-checkbox>\n  </div>\n  <div *ngIf=\"displayError\" class=\"mapgl-import-error\">\n    {{ errorMessage | translate}} <ng-container *ngIf=\"errorThreshold !== ''\"> - ({{'Max:' | translate }} {{errorThreshold}})</ng-container>\n  </div>\n</div>\n\n<div mat-dialog-actions class=\"mapgl-import-actions\">\n  <button mat-button (click)=\"onCancel()\">{{ 'Cancel' | translate}}</button>\n  <button mat-button [disabled]=\"((importType === SHP || importType === KML || importType === GEOJSON) && !currentFile) || isRunning || (importType === WKT && wktContent === '')\" (click)=\"import()\">{{ 'Import' | translate}}</button>\n  <mat-progress-spinner *ngIf=\"isRunning\" [color]=\"'accent'\" [diameter]=\"40\" [mode]=\"'indeterminate'\">\n  </mat-progress-spinner>\n</div>\n", styles: ["@charset \"UTF-8\";::ng-deep .map-import-dialog{--mdc-dialog-subhead-size: 20px;--mdc-dialog-subhead-tracking: 0;--mat-dialog-with-actions-content-padding: 0}::ng-deep .map-import-dialog .mdc-dialog__surface{padding:12px 24px}::ng-deep .map-import-dialog .mat-mdc-dialog-content{padding:0}.mapgl-import-body{min-width:500px;display:flex;flex-direction:column}.mapgl-import-zoom{display:flex;align-content:center;align-items:center;height:60px}.mapgl-import-error{color:#f44336;font-size:75%}.mapgl-import-upload{padding-left:10px;padding-top:10px}.mapgl-import-upload:focus{outline:none}.mapgl-import-radio-button{padding:10px}.mapgl-import-text-area{box-sizing:border-box;width:100%;height:200px;resize:vertical}\n"], dependencies: [{ kind: "directive", type: i2.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { kind: "directive", type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "directive", type: i3.DefaultValueAccessor, selector: "input:not([type=checkbox])[formControlName],textarea[formControlName],input:not([type=checkbox])[formControl],textarea[formControl],input:not([type=checkbox])[ngModel],textarea[ngModel],[ngDefaultControl]" }, { kind: "directive", type: i3.NgControlStatus, selector: "[formControlName],[ngModel],[formControl]" }, { kind: "directive", type: i3.NgModel, selector: "[ngModel]:not([formControlName]):not([formControl])", inputs: ["name", "disabled", "ngModel", "ngModelOptions"], outputs: ["ngModelChange"], exportAs: ["ngModel"] }, { kind: "component", type: i4.MatButton, selector: "    button[mat-button], button[mat-raised-button], button[mat-flat-button],    button[mat-stroked-button]  ", exportAs: ["matButton"] }, { kind: "component", type: i5.MatCheckbox, selector: "mat-checkbox", inputs: ["aria-label", "aria-labelledby", "aria-describedby", "id", "required", "labelPosition", "name", "value", "disableRipple", "tabIndex", "color", "disabledInteractive", "checked", "disabled", "indeterminate"], outputs: ["change", "indeterminateChange"], exportAs: ["matCheckbox"] }, { kind: "directive", type: i1.MatDialogTitle, selector: "[mat-dialog-title], [matDialogTitle]", inputs: ["id"], exportAs: ["matDialogTitle"] }, { kind: "directive", type: i1.MatDialogActions, selector: "[mat-dialog-actions], mat-dialog-actions, [matDialogActions]", inputs: ["align"] }, { kind: "directive", type: i1.MatDialogContent, selector: "[mat-dialog-content], mat-dialog-content, [matDialogContent]" }, { kind: "component", type: i6.MatProgressSpinner, selector: "mat-progress-spinner, mat-spinner", inputs: ["color", "mode", "value", "diameter", "strokeWidth"], exportAs: ["matProgressSpinner"] }, { kind: "directive", type: i7.MatRadioGroup, selector: "mat-radio-group", inputs: ["color", "name", "labelPosition", "value", "selected", "disabled", "required", "disabledInteractive"], outputs: ["change"], exportAs: ["matRadioGroup"] }, { kind: "component", type: i7.MatRadioButton, selector: "mat-radio-button", inputs: ["id", "name", "aria-label", "aria-labelledby", "aria-describedby", "disableRipple", "tabIndex", "checked", "value", "labelPosition", "disabled", "required", "color", "disabledInteractive"], outputs: ["change"], exportAs: ["matRadioButton"] }, { kind: "pipe", type: i8.TranslatePipe, name: "translate" }] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: MapImportDialogComponent, decorators: [{
            type: Component,
            args: [{ selector: 'arlas-map-import-dialog', template: "<h1 mat-dialog-title class=\"mapgl-import-title\">{{ 'Import file' | translate}}</h1>\n<div mat-dialog-content class=\"mapgl-import-body\">\n  <mat-radio-group class=\"mapgl-import-radio-group\" [(ngModel)]=\"importType\" (change)=\"changeType()\">\n    <mat-radio-button *ngFor=\"let type of allowedImportType\" [value]=\"type\" class=\"mapgl-import-radio-button\">\n      {{ type | translate }}\n    </mat-radio-button>\n  </mat-radio-group>\n\n  <input *ngIf=\"importType !== WKT\" type=\"file\" class=\"mapgl-import-upload\" #fileInput accept=\"{{allowedFileExtension}}\" (change)=\"onFileChange($event.target.files)\">\n  <textarea *ngIf=\"importType === WKT\" class=\"mapgl-import-text-area\" [(ngModel)]=\"wktContent\" placeholder=\"{{'Paste your WKT' | translate}}\" (change)=\"onTextChange()\"></textarea>\n\n  <div class=\"mapgl-import-zoom\">\n    <mat-checkbox [(ngModel)]=\"fitResult\">{{ 'Zoom to result' | translate}}</mat-checkbox>\n  </div>\n  <div *ngIf=\"displayError\" class=\"mapgl-import-error\">\n    {{ errorMessage | translate}} <ng-container *ngIf=\"errorThreshold !== ''\"> - ({{'Max:' | translate }} {{errorThreshold}})</ng-container>\n  </div>\n</div>\n\n<div mat-dialog-actions class=\"mapgl-import-actions\">\n  <button mat-button (click)=\"onCancel()\">{{ 'Cancel' | translate}}</button>\n  <button mat-button [disabled]=\"((importType === SHP || importType === KML || importType === GEOJSON) && !currentFile) || isRunning || (importType === WKT && wktContent === '')\" (click)=\"import()\">{{ 'Import' | translate}}</button>\n  <mat-progress-spinner *ngIf=\"isRunning\" [color]=\"'accent'\" [diameter]=\"40\" [mode]=\"'indeterminate'\">\n  </mat-progress-spinner>\n</div>\n", styles: ["@charset \"UTF-8\";::ng-deep .map-import-dialog{--mdc-dialog-subhead-size: 20px;--mdc-dialog-subhead-tracking: 0;--mat-dialog-with-actions-content-padding: 0}::ng-deep .map-import-dialog .mdc-dialog__surface{padding:12px 24px}::ng-deep .map-import-dialog .mat-mdc-dialog-content{padding:0}.mapgl-import-body{min-width:500px;display:flex;flex-direction:column}.mapgl-import-zoom{display:flex;align-content:center;align-items:center;height:60px}.mapgl-import-error{color:#f44336;font-size:75%}.mapgl-import-upload{padding-left:10px;padding-top:10px}.mapgl-import-upload:focus{outline:none}.mapgl-import-radio-button{padding:10px}.mapgl-import-text-area{box-sizing:border-box;width:100%;height:200px;resize:vertical}\n"] }]
        }], ctorParameters: () => [{ type: i1.MatDialogRef }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [MAT_DIALOG_DATA]
                }] }], propDecorators: { file: [{
                type: Output
            }], importRun: [{
                type: Output
            }], fileInput: [{
                type: ViewChild,
                args: ['fileInput', { static: false }]
            }] } });
const SIMPLE_GEOMETRY_OBJECT = ['Polygon', 'Point', 'LineString'];
/** L: a layer class/interface.
 *  S: a source class/interface.
 *  M: a Map configuration class/interface.
 */
export class MapImportComponent {
    constructor(dialog, mapService) {
        this.dialog = dialog;
        this.mapService = mapService;
        this.SHP = 'shp';
        this.KML = 'kml';
        this.WKT = 'wkt';
        this.GEOJSON = 'geojson';
        this.SELF_INTERSECT = marker('Geometry is not valid due to self-intersection');
        this.PARSING_ISSUE = marker('Problem parsing input file');
        this.FILE_TOO_LARGE = marker('File is too large');
        this.GEOMETRY_INVALID = marker('Geometry is not valid');
        this.TOO_MANY_VERTICES = marker('Too many vertices in a polygon');
        this.TOO_MANY_FEATURES = marker('Too many features');
        this.TIMEOUT = marker('Timeout');
        this.tooManyVertex = false;
        this.fitResult = false;
        this.SOURCE_NAME_POLYGON_LABEL = 'polygon_label';
        this.emptyData = {
            'type': 'FeatureCollection',
            'features': []
        };
        this.featureIndex = 0;
        this.maxLoadingTime = 20000;
        this.allowedImportType = [this.SHP, this.KML, this.WKT, this.GEOJSON];
        this.allowedGeometryObjectType = ['Polygon'];
        this.imported = new Subject();
        this.error = new Subject();
    }
    promiseTimeout(ms, promise) {
        // Create a promise that rejects in <ms> milliseconds
        const timeout = new Promise((resolve, reject) => {
            const id = setTimeout(() => {
                clearTimeout(id);
                reject(new Error(this.TIMEOUT));
            }, ms);
        });
        // Returns a race between our timeout and the passed in promise
        return Promise.race([
            promise,
            timeout
        ]);
    }
    buildAllowedGeometryForImportType(importType) {
        this._currentAllowedGeom = [];
        this.allowedGeometryObjectType.forEach(allowed => {
            this._currentAllowedGeom = this._currentAllowedGeom.concat(this.getAllowedGeom(allowed));
        });
        if (importType === this.KML) {
            this._currentAllowedGeom.push('GeometryCollection', 'MultiGeometry');
        }
        else if (importType === this.WKT) {
            this._currentAllowedGeom.push('GeometryCollection');
        }
    }
    getAllowedGeom(allowed) {
        if (allowed === 'Polygon') {
            return ['Polygon', 'MultiPolygon'];
        }
        if (allowed === 'Point') {
            return ['Point', 'MultiPoint'];
        }
    }
    openDialog(defaultFitResult) {
        this.dialogRef = this.dialog.open(MapImportDialogComponent, { data: { allowedImportType: this.allowedImportType, defaultFitResult }, panelClass: 'map-import-dialog' });
        this.dialogRef.componentInstance.file.subscribe((file) => {
            this.currentFile = file;
        });
        this.dialogRef.componentInstance.importRun.subscribe(importOptions => {
            this.fitResult = importOptions.fitResult;
            this.buildAllowedGeometryForImportType(importOptions.type);
            this.import(importOptions.type, importOptions.wktContent);
        });
    }
    import(importType, content) {
        this.dialogRef.componentInstance.isRunning = true;
        this.tooManyVertex = false;
        this.jszip = new JSZip();
        let processPromise;
        if (importType === this.SHP) {
            processPromise = this.processAllShape();
        }
        else if (importType === this.KML) {
            processPromise = this.processAllKml();
        }
        else if (importType === this.WKT) {
            processPromise = this.processWKT(content);
        }
        else if (importType === this.GEOJSON) {
            processPromise = this.processJson();
        }
        this.promiseTimeout(this.maxLoadingTime, processPromise).catch(error => {
            if (importType !== this.WKT) {
                this.reader.abort();
            }
            this.throwError(error);
        });
    }
    buildFeature(geom, feature, geometryType, bbox) {
        const f = {
            type: 'Feature',
            geometry: {
                coordinates: geom,
                type: geometryType ?? geom.type
            },
            properties: feature.properties
        };
        if (bbox) {
            f.geometry['bbox'] = feature.geometry.bbox;
        }
        return f;
    }
    handleSimpleGeometry(feature, centroids, importedGeojson) {
        // avoid self intersect control for point
        if (feature.geometry.type === 'Point' || gpsi(feature).geometry.coordinates.length === 0) {
            this.addFeature(feature, centroids, importedGeojson, ++this.featureIndex);
        }
        else {
            throw new Error('Geometry is not valid due to self-intersection');
        }
    }
    handleMultiGeometry(feature, centroids, importedGeojson) {
        // Create a new Polygon feature for each polygon in the MultiPolygon
        // All properties of the MultiPolygon are copied in each feature created
        const geomType = (feature.geometry.type === 'MultiPolygon') ? 'Polygon' : 'Point';
        feature.geometry.coordinates.forEach(geom => {
            const newFeature = this.buildFeature(geom, feature, geomType, true);
            this.handleSimpleGeometry(newFeature, centroids, importedGeojson);
        });
    }
    handleGeometryCollection(feature, centroids, importedGeojson) {
        // Create a new Polygon feature for each polygon in the MultiPolygon
        // All properties of the MultiPolygon are copied in each feature created
        const simpleGeometry = this._currentAllowedGeom.filter(g => SIMPLE_GEOMETRY_OBJECT.includes(g));
        feature.geometry.geometries.filter(geom => simpleGeometry.includes(geom.type)).forEach(geom => {
            const newFeature = this.buildFeature(geom, feature);
            this.handleSimpleGeometry(newFeature, centroids, importedGeojson);
        });
    }
    handleFeatureCollection(feature, centroids, importedGeojson) {
        feature.features.filter(feature => this._currentAllowedGeom.includes(feature.geometry.type))
            .forEach((feature) => {
            const multiGeometry = this._currentAllowedGeom.filter(g => !SIMPLE_GEOMETRY_OBJECT.includes(g));
            if (multiGeometry.includes(feature.geometry.type)) {
                this.handleMultiGeometry(feature, centroids, importedGeojson);
            }
            else {
                this.handleSimpleGeometry(feature, centroids, importedGeojson);
            }
        });
    }
    /** *************/
    /** *** KML *****/
    /** *************/
    readKmlFile() {
        return new Promise((resolve, reject) => {
            this.reader = new FileReader();
            const reader = this.reader;
            reader.onload = () => {
                resolve(reader.result);
            };
            reader.onerror = () => {
                reader.abort();
                reject(new Error(this.PARSING_ISSUE));
            };
            if (this.maxFileSize && this.currentFile.size > this.maxFileSize) {
                reject(new Error(this.FILE_TOO_LARGE));
            }
            else if (this.currentFile.name.split('.').pop().toLowerCase() === this.KML) {
                reader.readAsText(this.currentFile);
            }
            else if (this.currentFile.name.split('.').pop().toLowerCase() === 'kmz') {
                reader.readAsArrayBuffer(this.currentFile);
            }
            else {
                reject(new Error(marker('Only `kml` or `zip` file is allowed')));
            }
        });
    }
    resolveFileFromGzip(result, resolve) {
        this.jszip.loadAsync(result).then(kmzContent => {
            const kmlFile = Object.keys(kmzContent.files).filter(file => file.split('.').pop().toLowerCase() === this.KML)[0];
            this.jszip.file(kmlFile).async('text').then((data) => resolve(data));
        });
    }
    processAllKml() {
        const readKmlFile = this.readKmlFile();
        let readKmzFile = readKmlFile;
        if (this.currentFile.name.split('.').pop().toLowerCase() === 'kmz') {
            readKmzFile = readKmlFile.then(result => new Promise((resolve, reject) => {
                this.resolveFileFromGzip(result, resolve);
            }));
        }
        const parseKml = readKmzFile.then((file) => new Promise((resolve, reject) => {
            const geojson = toGeoJSON.kml((new DOMParser()).parseFromString(file, 'text/xml'));
            resolve(geojson);
        }));
        const geojsonParserPromise = parseKml.then((geojson) => new Promise((resolve, reject) => {
            this.computeGeojson(geojson, reject, resolve);
        }));
        return Promise.all([geojsonParserPromise])
            .then(([importedResult]) => {
            this.setImportedData(importedResult);
        });
    }
    /** *************/
    /** * GEOJSON ***/
    /** *************/
    readJsonFile() {
        return new Promise((resolve, reject) => {
            this.reader = new FileReader();
            const reader = this.reader;
            reader.onload = () => {
                resolve(reader.result);
            };
            reader.onerror = () => {
                reader.abort();
                reject(new Error(this.PARSING_ISSUE));
            };
            if (this.maxFileSize && this.currentFile.size > this.maxFileSize) {
                reject(new Error(this.FILE_TOO_LARGE));
            }
            else {
                const extension = this.currentFile.name.split('.').pop().toLowerCase();
                if (extension === 'json' || extension === 'geojson') {
                    reader.readAsText(this.currentFile);
                }
                else {
                    reject(new Error(marker('Only `json` or `geojson` file is allowed')));
                }
            }
        });
    }
    processJson() {
        const readJsonFile = this.readJsonFile();
        const parseJson = readJsonFile.then((fileContent) => new Promise((resolve, reject) => {
            const feature = JSON.parse(fileContent);
            if (valid(feature) && (this._currentAllowedGeom.includes(feature.geometry) || feature.type === 'FeatureCollection')) {
                const centroides = new Array();
                const importedGeojson = {
                    type: 'FeatureCollection',
                    features: []
                };
                try {
                    if (feature.geometry && (feature.geometry.type === 'Polygon' || feature.geometry.type === 'Point')) {
                        this.handleSimpleGeometry(feature, centroides, importedGeojson);
                    }
                    else if (feature.geometry && (feature.geometry.type === 'MultiPolygon' || feature.geometry.type === 'MultiPoint')) {
                        this.handleMultiGeometry(feature, centroides, importedGeojson);
                    }
                    else if (feature.type && feature.type === 'FeatureCollection') {
                        this.handleFeatureCollection(feature, centroides, importedGeojson);
                    }
                    resolve({ geojson: importedGeojson, centroides: centroides });
                }
                catch (e) {
                    reject(e);
                }
            }
            else {
                reject(new Error(this.GEOMETRY_INVALID));
            }
        }));
        return Promise.all([parseJson])
            .then(([importedResult]) => {
            this.setImportedData(importedResult);
        });
    }
    /** *************/
    /** ** SHAPE ****/
    /** *************/
    readZipFile() {
        return new Promise((resolve, reject) => {
            this.reader = new FileReader();
            const reader = this.reader;
            reader.onload = () => {
                const resultToArray = new Uint8Array(reader.result);
                if (resultToArray.length === 0) {
                    reader.abort();
                    reject(new Error(marker('File is empty')));
                }
                else {
                    resolve(reader.result);
                }
            };
            reader.onerror = () => {
                reader.abort();
                reject(new Error(this.PARSING_ISSUE));
            };
            if (this.maxFileSize && this.currentFile.size > this.maxFileSize) {
                reject(new Error(this.FILE_TOO_LARGE));
            }
            else if (this.currentFile.name.split('.').pop().toLowerCase() !== 'zip') {
                reject(new Error(marker('Only `zip` file is allowed')));
            }
            else {
                reader.readAsArrayBuffer(this.currentFile);
            }
        });
    }
    areFilesInvalid(zipResult) {
        const testArray = Object.keys(zipResult.files).map(fileName => fileName.split('.').pop().toLowerCase());
        return (testArray.filter(elem => elem === this.SHP || elem === 'shx' || elem === 'dbf').length < 3) &&
            (testArray.filter(elem => elem === 'json').length !== 1);
    }
    processAllShape() {
        const fileReaderPromise = this.readZipFile();
        const zipLoaderPromise = fileReaderPromise.then((buffer) => new Promise((resolve, reject) => {
            this.jszip.loadAsync(buffer).then(zipResult => {
                if (this.areFilesInvalid(zipResult)) {
                    reject(new Error(marker('Zip file must contain at least a `*.shp`, `*.shx` and `*.dbf` or a `*.json`')));
                }
                else {
                    resolve(buffer);
                }
            });
        }));
        const shapeParserPromise = zipLoaderPromise
            .then(buffer => shp(buffer));
        const geojsonParserPromise = shapeParserPromise.then(geojson => new Promise((resolve, reject) => {
            this.computeGeojson(geojson, reject, resolve);
        }));
        return Promise.all([fileReaderPromise, zipLoaderPromise, shapeParserPromise, geojsonParserPromise])
            .then(([a, b, c, importedResult]) => {
            this.setImportedData(importedResult);
        });
    }
    /** *************/
    /** **  WKT  ****/
    /** *************/
    processWKT(wkt) {
        const wktParserPromise = new Promise((resolve, reject) => {
            const geojsonWKT = parse(wkt);
            const centroides = new Array();
            const importedGeojson = {
                type: 'FeatureCollection',
                features: []
            };
            if (geojsonWKT && valid(geojsonWKT) && this._currentAllowedGeom.includes(geojsonWKT.type)) {
                const feature = {
                    type: 'Feature',
                    geometry: geojsonWKT,
                    properties: { arlas_id: null }
                };
                this.handleGeom(feature, centroides, importedGeojson, reject);
                resolve({ geojson: importedGeojson, centroides: centroides });
            }
            else {
                reject(new Error(this.GEOMETRY_INVALID));
            }
        });
        return Promise.all([wktParserPromise]).then(([importedResult]) => {
            this.setImportedData(importedResult);
        });
    }
    /** *************/
    /** ** TOOLS ****/
    /** *************/
    clearPolygons() {
        // Clean source of imported polygons
        const labelSource = this.mapService.getSource(this.SOURCE_NAME_POLYGON_LABEL, this.mapComponent.map);
        this.featureIndex = 0;
        this.mapComponent.onAoiChanged.next(this.emptyData);
        if (labelSource !== undefined) {
            this.mapService.setDataToGeojsonSource(labelSource, this.emptyData);
        }
    }
    addFeature(feature, centroides, importedGeojson, index) {
        feature.properties.arlas_id = index;
        const cent = this.calcCentroid(feature);
        centroides.push(cent);
        importedGeojson.features.push(feature);
    }
    setImportedData(importedResult) {
        if (this.tooManyVertex) {
            throw new Error(this.TOO_MANY_VERTICES);
        }
        else if (this.maxFeatures && importedResult.geojson.features.length > this.maxFeatures) {
            throw new Error(this.TOO_MANY_FEATURES);
        }
        else if (importedResult.geojson.features.length > 0) {
            this.dialogRef.componentInstance.isRunning = false;
            if (this.fitResult) {
                this.mapComponent.fitToPaddedBounds(extent(importedResult.geojson));
            }
            if (this.mapComponent.drawData.features.length > 0) {
                this.mapComponent.drawData.features.forEach(df => importedResult.geojson.features.push(df));
            }
            this.mapComponent.drawComponent.draw.changeMode('static');
            this.imported.next(importedResult.geojson.features);
            this.mapComponent.onAoiChanged.next(importedResult.geojson);
            this.dialogRef.close();
        }
        else {
            throw new Error(marker('No polygon to display in this file'));
        }
    }
    calcCentroid(feature) {
        let cent;
        if (feature.type === 'Point') {
            cent = centroid(feature);
        }
        else {
            if (!this.maxVertexByPolygon) {
                this.maxVertexByPolygon = 100;
            }
            if (this.maxVertexByPolygon && feature.geometry.coordinates[0].length - 1 > this.maxVertexByPolygon) {
                this.tooManyVertex = true;
            }
            const poly = polygon(feature.geometry.coordinates);
            cent = centroid(poly);
        }
        cent.properties.arlas_id = feature.properties.arlas_id;
        return cent;
    }
    throwError(error) {
        this.dialogRef.componentInstance.displayError = true;
        this.dialogRef.componentInstance.isRunning = false;
        this.dialogRef.componentInstance.errorMessage = error.message;
        switch (this.dialogRef.componentInstance.errorMessage) {
            case this.TOO_MANY_FEATURES:
                this.dialogRef.componentInstance.errorThreshold = this.maxFeatures.toString();
                break;
            case this.TOO_MANY_VERTICES:
                this.dialogRef.componentInstance.errorThreshold = this.maxVertexByPolygon.toString();
                break;
            case this.FILE_TOO_LARGE:
                this.dialogRef.componentInstance.errorThreshold = this.formatBytes(this.maxFileSize);
                break;
            case this.TIMEOUT:
                this.dialogRef.componentInstance.errorThreshold = this.maxLoadingTime + ' ms';
                break;
            default:
                this.dialogRef.componentInstance.errorThreshold = '';
        }
        if (this.dialogRef.componentInstance.fileInput) {
            this.dialogRef.componentInstance.fileInput.nativeElement.value = '';
        }
        this.dialogRef.componentInstance.currentFile = null;
        this.error.next(error.message);
    }
    formatBytes(bytes, decimals = 2) {
        if (bytes === 0) {
            return '0 Bytes';
        }
        const k = 1024;
        const dm = decimals < 0 ? 0 : decimals;
        const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];
        const i = Math.floor(Math.log(bytes) / Math.log(k));
        return parseFloat((bytes / Math.pow(k, i)).toFixed(dm)) + ' ' + sizes[i];
    }
    computeGeojson(geojson, reject, resolve) {
        if (valid(geojson)) {
            const centroides = new Array();
            const importedGeojson = {
                type: 'FeatureCollection',
                features: []
            };
            geojson.features.filter(feature => this._currentAllowedGeom.includes(feature.geometry.type))
                .forEach((feature) => {
                this.handleGeom(feature, centroides, importedGeojson, reject);
            });
            resolve({ geojson: importedGeojson, centroides: centroides });
        }
        else {
            reject(new Error('Geometry is not valid'));
        }
    }
    handleGeom(feature, centroides, importedGeojson, reject) {
        try {
            if (feature.geometry.type === 'GeometryCollection' || feature.geometry.type === 'MultiGeometry') {
                // Create a new Polygon feature for each polygon in the MultiPolygon
                // All properties of the MultiPolygon are copied in each feature created
                this.handleGeometryCollection(feature, centroides, importedGeojson);
            }
            else if (feature.geometry.type === 'MultiPolygon' || feature.geometry.type === 'MultiPoint') {
                this.handleMultiGeometry(feature, centroides, importedGeojson);
            }
            else {
                this.handleSimpleGeometry(feature, centroides, importedGeojson);
            }
        }
        catch (e) {
            reject(new Error('Error during import'));
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: MapImportComponent, deps: [{ token: i1.MatDialog }, { token: i9.ArlasMapFrameworkService }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "18.2.13", type: MapImportComponent, selector: "arlas-map-import", inputs: { mapComponent: "mapComponent", maxVertexByPolygon: "maxVertexByPolygon", maxFeatures: "maxFeatures", maxFileSize: "maxFileSize", maxLoadingTime: "maxLoadingTime", allowedImportType: "allowedImportType", allowedGeometryObjectType: "allowedGeometryObjectType" }, outputs: { imported: "imported", error: "error" }, ngImport: i0, template: "\n\n", styles: ["@charset \"UTF-8\";\n"] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: MapImportComponent, decorators: [{
            type: Component,
            args: [{ selector: 'arlas-map-import', template: "\n\n", styles: ["@charset \"UTF-8\";\n"] }]
        }], ctorParameters: () => [{ type: i1.MatDialog }, { type: i9.ArlasMapFrameworkService }], propDecorators: { mapComponent: [{
                type: Input
            }], maxVertexByPolygon: [{
                type: Input
            }], maxFeatures: [{
                type: Input
            }], maxFileSize: [{
                type: Input
            }], maxLoadingTime: [{
                type: Input
            }], allowedImportType: [{
                type: Input
            }], allowedGeometryObjectType: [{
                type: Input
            }], imported: [{
                type: Output
            }], error: [{
                type: Output
            }] } });
