// arlas-map/esm2022/lib/coordinates/coordinates.module.mjs
/*
 * Licensed to Gisaïa under one or more contributor
 * license agreements. See the NOTICE.txt file distributed with
 * this work for additional information regarding copyright
 * ownership. Gisaïa licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { CommonModule } from '@angular/common';
import { MatIconModule } from '@angular/material/icon';
import { TranslateModule } from '@ngx-translate/core';
import { NgModule } from '@angular/core';
import { CoordinatesComponent } from './coordinates.component';
import { MatFormFieldModule } from '@angular/material/form-field';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { MatTooltipModule } from '@angular/material/tooltip';
import { MatInputModule } from '@angular/material/input';
import { CoordinatesErrorPipe } from './coordinates.pipe';
import * as i0 from "@angular/core";
export class CoordinatesModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: CoordinatesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.13", ngImport: i0, type: CoordinatesModule, declarations: [CoordinatesComponent, CoordinatesErrorPipe], imports: [CommonModule,
            MatIconModule,
            MatFormFieldModule,
            FormsModule,
            ReactiveFormsModule,
            MatTooltipModule,
            MatInputModule,
            TranslateModule], exports: [CoordinatesComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: CoordinatesModule, imports: [CommonModule,
            MatIconModule,
            MatFormFieldModule,
            FormsModule,
            ReactiveFormsModule,
            MatTooltipModule,
            MatInputModule,
            TranslateModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: CoordinatesModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        MatIconModule,
                        MatFormFieldModule,
                        FormsModule,
                        ReactiveFormsModule,
                        MatTooltipModule,
                        MatInputModule,
                        TranslateModule
                    ],
                    declarations: [CoordinatesComponent, CoordinatesErrorPipe],
                    exports: [CoordinatesComponent]
                }]
        }] });
