diff --git a/node_modules/maplibre-gl/dist/maplibre-gl.d.ts b/node_modules/maplibre-gl/dist/maplibre-gl.d.ts
index 9f6c16c..5bb0086 100644
--- a/node_modules/maplibre-gl/dist/maplibre-gl.d.ts
+++ b/node_modules/maplibre-gl/dist/maplibre-gl.d.ts
@@ -3,7 +3,7 @@
 import Point from '@mapbox/point-geometry';
 import TinySDF from '@mapbox/tiny-sdf';
 import { GeoJSONVTOptions } from '@maplibre/geojson-vt';
-import { Color, ColorArray, CompositeExpression, DiffCommand, DiffOperations, Feature, Feature as StyleFeature, FeatureFilter, FeatureState, FilterSpecification, Formatted, FormattedSection, GeoJSONSourceSpecification, GlobalProperties, ICanonicalTileID, IMercatorCoordinate, ImageSourceSpecification, InterpolationType, LayerSpecification, LightSpecification, NumberArray, Padding, ProjectionSpecification, PromoteIdSpecification, PropertyValueSpecification, RasterDEMSourceSpecification, RasterSourceSpecification, ResolvedImage, SkySpecification, SourceExpression, SourceSpecification, SpriteSpecification, StateSpecification, StylePropertyExpression, StylePropertySpecification, StyleSpecification, TerrainSpecification, TransitionSpecification, VariableAnchorOffsetCollection, VectorSourceSpecification, VideoSourceSpecification, VisibilityExpression, VisibilitySpecification } from '@maplibre/maplibre-gl-style-spec';
+import { Color, ColorArray, CompositeExpression, DiffCommand, DiffOperations, Feature, FeatureFilter, FeatureState, FilterSpecification, Formatted, FormattedSection, GeoJSONSourceSpecification, GlobalProperties, ICanonicalTileID, IMercatorCoordinate, ImageSourceSpecification, InterpolationType, LayerSpecification, LightSpecification, NumberArray, Padding, ProjectionSpecification, PromoteIdSpecification, PropertyValueSpecification, RasterDEMSourceSpecification, RasterSourceSpecification, ResolvedImage, SkySpecification, SourceExpression, SourceSpecification, SpriteSpecification, StateSpecification, Feature as StyleFeature, StylePropertyExpression, StylePropertySpecification, StyleSpecification, TerrainSpecification, TransitionSpecification, VariableAnchorOffsetCollection, VectorSourceSpecification, VideoSourceSpecification, VisibilityExpression, VisibilitySpecification } from '@maplibre/maplibre-gl-style-spec';
 import { VectorTileFeatureLike, VectorTileLayerLike } from '@maplibre/vt-pbf';
 import { mat2, mat4, vec3, vec4 } from 'gl-matrix';
 import KDBush from 'kdbush';
@@ -15126,6 +15126,7 @@ export declare function importScriptInWorkers(workerUrl: string): Promise<void[]
 export type * from "@maplibre/maplibre-gl-style-spec";
 
 export {
+	CirclePaintProps,
 	Color,
 	ColorArray,
 	CompositeExpression,
@@ -15136,17 +15137,20 @@ export {
 	Feature,
 	FeatureFilter,
 	FeatureState,
+	FillPaintProps,
 	FilterSpecification,
 	Formatted,
 	FormattedSection,
 	GeoJSONSourceSpecification,
 	GlobalProperties,
+	HeatmapPaintProps,
 	ICanonicalTileID,
 	IMercatorCoordinate,
 	ImageSourceSpecification,
 	InterpolationType,
 	LayerSpecification,
 	LightSpecification,
+	LinePaintProps,
 	Map$1 as Map,
 	Map$1 as MapLibreMap,
 	NumberArray,
@@ -15166,15 +15170,18 @@ export {
 	StylePropertyExpression,
 	StylePropertySpecification,
 	StyleSpecification,
+	SymbolLayoutProps,
+	SymbolPaintProps,
 	TerrainSpecification,
 	TransitionSpecification,
 	VariableAnchorOffsetCollection,
 	VectorSourceSpecification,
 	VideoSourceSpecification,
 	VisibilityExpression,
-	VisibilitySpecification,
+	VisibilitySpecification
 };
 
 export as namespace maplibregl;
 
-export {};
+export { };
+
