// arlas-web-components/esm2022/lib/components/results/model/column.mjs
/*
 * Licensed to Gisaïa under one or more contributor
 * license agreements. See the NOTICE.txt file distributed with
 * this work for additional information regarding copyright
 * ownership. Gisaïa licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { SortEnum } from '../utils/enumerations/sortEnum';
export class Column {
    constructor(columnName, fieldName, dataType) {
        /**
         * @description Sort direction to apply to th column : ascending, descending or none
         */
        this.sortDirection = SortEnum.none;
        /**
         * @description Whether this column represents an id field.
         */
        this.isIdField = false;
        /**
         * @description Whether the cells of this column contains a toggle button.
         */
        this.isToggleField = false;
        /**
         * @description Whether the filter search column has a dropdown.
         */
        this.dropdown = false;
        /**
         * @description Size of the dropdown list.
         */
        this.dropdownsize = 10;
        /**
         * @description Whether to allow colorizing cells that are within this column
         */
        this.useColorService = false;
        this.columnName = columnName;
        this.fieldName = fieldName;
        this.dataType = dataType;
    }
}
