// arlas-web-components/esm2022/lib/components/results/model/itemComponent.mjs
/*
 * Licensed to Gisaïa under one or more contributor
 * license agreements. See the NOTICE.txt file distributed with
 * this work for additional information regarding copyright
 * ownership. Gisaïa licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
export class ItemComponent {
    setSelectedItem(isChecked, identifier, selectedItems) {
        isChecked = !isChecked;
        if (isChecked) {
            if (!selectedItems.has(identifier)) {
                selectedItems.add(identifier);
            }
        }
        else {
            if (selectedItems.has(identifier)) {
                selectedItems.delete(identifier);
            }
        }
    }
    retrieveAdditionalInfo(detailedDataRetriever, item) {
        if (detailedDataRetriever !== null && item.itemDetailedData.length === 0) {
            this.retrievedDataEvent = detailedDataRetriever.getData(((String)(item.identifier)));
            this.retrievedDataEvent.subscribe(additionalInfo => {
                item.actions = new Array();
                additionalInfo.actions.forEach(action => {
                    item.actions.push({
                        id: action.id,
                        label: action.label,
                        actionBus: action.actionBus,
                        cssClass: action.cssClass,
                        tooltip: action.tooltip,
                        reverseAction: action.reverseAction,
                        icon: action.icon,
                        fields: action.fields,
                        show: action.show
                    });
                });
                additionalInfo.details.forEach((v, k) => {
                    const details = new Array();
                    v.forEach((value, key) => details.push({ key: key, value: value }));
                    item.itemDetailedData.push({ group: k, details: details });
                });
                if (additionalInfo.attachments) {
                    item.attachments = new Array();
                    additionalInfo.attachments.forEach(attachment => {
                        item.attachments.push({
                            label: attachment.label,
                            url: attachment.url,
                            type: attachment.type,
                            description: attachment.description,
                            icon: attachment.icon
                        });
                    });
                }
            });
        }
    }
}
