// arlas-web-components/esm2022/lib/pipes/get-color/get-color.pipe.mjs
/*
 * Licensed to Gisaïa under one or more contributor
 * license agreements. See the NOTICE.txt file distributed with
 * this work for additional information regarding copyright
 * ownership. Gisaïa licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { Injectable, Pipe } from '@angular/core';
import { ArlasColorService } from '../../services/color.generator.service';
import * as i0 from "@angular/core";
import * as i1 from "../../services/color.generator.service";
export class GetColorPipe {
    constructor(colorService) {
        this.colorService = colorService;
    }
    transform(key) {
        return this.colorService.getColor(key);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: GetColorPipe, deps: [{ token: i1.ArlasColorService }], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.2.13", ngImport: i0, type: GetColorPipe, name: "getColor" }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: GetColorPipe }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: GetColorPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'getColor'
                }]
        }, {
            type: Injectable
        }], ctorParameters: () => [{ type: i1.ArlasColorService }] });
