// arlas-map/esm2022/lib/map-import/map-import.module.mjs
/*
 * Licensed to Gisaïa under one or more contributor
 * license agreements. See the NOTICE.txt file distributed with
 * this work for additional information regarding copyright
 * ownership. Gisaïa licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { MatButtonModule } from '@angular/material/button';
import { MatCheckboxModule } from '@angular/material/checkbox';
import { MatDialogModule } from '@angular/material/dialog';
import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
import { MatRadioModule } from '@angular/material/radio';
import { MatSelectModule } from '@angular/material/select';
import { TranslateModule } from '@ngx-translate/core';
import { MapImportComponent, MapImportDialogComponent } from './map-import.component';
import * as i0 from "@angular/core";
export class MapImportModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: MapImportModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.13", ngImport: i0, type: MapImportModule, declarations: [MapImportComponent, MapImportDialogComponent], imports: [CommonModule,
            FormsModule,
            MatButtonModule,
            MatCheckboxModule,
            MatDialogModule,
            MatProgressSpinnerModule,
            MatSelectModule,
            MatRadioModule,
            TranslateModule], exports: [MapImportComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: MapImportModule, imports: [CommonModule,
            FormsModule,
            MatButtonModule,
            MatCheckboxModule,
            MatDialogModule,
            MatProgressSpinnerModule,
            MatSelectModule,
            MatRadioModule,
            TranslateModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: MapImportModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        MatButtonModule,
                        MatCheckboxModule,
                        MatDialogModule,
                        MatProgressSpinnerModule,
                        MatSelectModule,
                        MatRadioModule,
                        TranslateModule
                    ],
                    declarations: [MapImportComponent, MapImportDialogComponent],
                    exports: [MapImportComponent]
                }]
        }] });
