// arlas-map/esm2022/lib/basemaps/basemap.component.mjs
/*
 * Licensed to Gisaïa under one or more contributor
 * license agreements. See the NOTICE.txt file distributed with
 * this work for additional information regarding copyright
 * ownership. Gisaïa licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Subject } from 'rxjs/internal/Subject';
import { AbstractArlasMapGL } from '../map/AbstractArlasMapGL';
import { BasemapService } from './basemap.service';
import { ArlasMapFrameworkService } from '../arlas-map-framework.service';
import { AbstractArlasMapService } from '../arlas-map.service';
import { takeUntil } from 'rxjs';
import * as i0 from "@angular/core";
import * as i1 from "./basemap.service";
import * as i2 from "../arlas-map.service";
import * as i3 from "../arlas-map-framework.service";
import * as i4 from "@angular/common";
import * as i5 from "@angular/material/icon";
import * as i6 from "@ngx-translate/core";
/** L: a layer class/interface.
 *  S: a source class/interface.
 *  M: a Map configuration class/interface.
 */
export class BasemapComponent {
    constructor(basemapService, mapService, mapFrameworkService) {
        this.basemapService = basemapService;
        this.mapService = mapService;
        this.mapFrameworkService = mapFrameworkService;
        this._onDestroy$ = new Subject();
        this.basemapChanged = new EventEmitter();
        this.blur = new Subject();
        this.showList = false;
        this.basemapService.basemapChanged$.pipe(takeUntil(this._onDestroy$)).subscribe(() => this.basemapChanged.emit());
    }
    ngOnInit() {
        this.initBasemaps();
    }
    initBasemaps() {
        this.basemaps = this.basemapService.basemaps;
        if (this.basemaps) {
            const styles = this.basemaps.styles();
            if (styles) {
                this.showList = styles.length > 0;
                styles.filter(bm => !bm.image).forEach(bm => {
                    if (bm.type !== 'protomap' && !!bm.url) {
                        const splitUrl = bm.url.split('/style.json?key=');
                        if (splitUrl.length === 2) {
                            bm.image = `${splitUrl[0]}/0/0/0.png?key=${splitUrl[1]}`;
                        }
                    }
                });
            }
        }
    }
    /** Removes the old basemap and set the new one that is given as a parameter
     * @param newBasemap: Basemap selected by the user
     */
    onChangeBasemap(newBasemap) {
        const selectedBasemap = this.basemaps.getSelected();
        if (selectedBasemap.type === 'protomap') {
            this.basemapService.removeProtomapBasemap(this.map);
        }
        this.setBaseMapStyle(newBasemap);
    }
    setBaseMapStyle(newBasemap) {
        if (this.map) {
            this.basemapService.setBasemap(this.basemaps.getSelected().styleFile, newBasemap, this.map, this.mapSources);
        }
    }
    ngOnDestroy() {
        this._onDestroy$.next(true);
        this._onDestroy$.complete();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: BasemapComponent, deps: [{ token: i1.BasemapService }, { token: i2.AbstractArlasMapService }, { token: i3.ArlasMapFrameworkService }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "18.2.13", type: BasemapComponent, selector: "arlas-basemap", inputs: { map: "map", mapSources: "mapSources" }, outputs: { basemapChanged: "basemapChanged", blur: "blur" }, ngImport: i0, template: "<div *ngIf=\"showList\" class=\"basemap-container\">\n  <ng-container *ngFor=\"let style of basemaps?._styles\">\n    <div *ngIf=\"!style.errored\" class=\"basemap\" [class.selected]=\"style.name === basemaps?._selectedStyle?.name\"\n    (click)=\"onChangeBasemap(style)\" (keyDown)=\"onChangeBasemap(style)\">\n      <div class=\"image\">\n        <img *ngIf=\"!!style?.image && style?.image !== ''\" src=\"{{style?.image}}\" (error)=\"style.image = null\" />\n        <div *ngIf=\"!style?.image || (!!style?.image && style?.image === '')\" class=\"no-image\">\n          <mat-icon>wallpaper</mat-icon>\n        </div>\n        <div class=\"name\">{{style.name | translate}}</div>\n      </div>\n    </div>\n  </ng-container>\n</div>\n", styles: ["@charset \"UTF-8\";.basemap-container{border-radius:6px;position:absolute;top:10px;right:45px;display:flex;flex-flow:row wrap;max-width:243px;gap:3px;overflow:hidden}.basemap-container .basemap{border:2px solid white;background-color:#fff;border-radius:6px;cursor:pointer}.basemap-container .basemap .image{height:75px;width:75px;position:relative}.basemap-container .basemap .image .name{font-size:10px;line-height:15px;text-align:center;position:absolute;width:100%;bottom:0;background-color:#f5f5f580;border-radius:0 0 6px 6px;text-shadow:1px 1px 2px white;text-overflow:ellipsis;white-space:nowrap;overflow:hidden}.basemap-container .basemap .image img{width:75px;border-radius:6px;image-rendering:auto}.basemap-container .basemap .image .no-image{width:75px;height:75px;background-color:#fff;text-align:center;padding-top:20px}.basemap-container .selected{border:2px solid #337ab7;background-color:#337ab7;color:#337ab7;font-weight:600}\n"], dependencies: [{ kind: "directive", type: i4.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { kind: "directive", type: i4.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "component", type: i5.MatIcon, selector: "mat-icon", inputs: ["color", "inline", "svgIcon", "fontSet", "fontIcon"], exportAs: ["matIcon"] }, { kind: "pipe", type: i6.TranslatePipe, name: "translate" }] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: BasemapComponent, decorators: [{
            type: Component,
            args: [{ selector: 'arlas-basemap', template: "<div *ngIf=\"showList\" class=\"basemap-container\">\n  <ng-container *ngFor=\"let style of basemaps?._styles\">\n    <div *ngIf=\"!style.errored\" class=\"basemap\" [class.selected]=\"style.name === basemaps?._selectedStyle?.name\"\n    (click)=\"onChangeBasemap(style)\" (keyDown)=\"onChangeBasemap(style)\">\n      <div class=\"image\">\n        <img *ngIf=\"!!style?.image && style?.image !== ''\" src=\"{{style?.image}}\" (error)=\"style.image = null\" />\n        <div *ngIf=\"!style?.image || (!!style?.image && style?.image === '')\" class=\"no-image\">\n          <mat-icon>wallpaper</mat-icon>\n        </div>\n        <div class=\"name\">{{style.name | translate}}</div>\n      </div>\n    </div>\n  </ng-container>\n</div>\n", styles: ["@charset \"UTF-8\";.basemap-container{border-radius:6px;position:absolute;top:10px;right:45px;display:flex;flex-flow:row wrap;max-width:243px;gap:3px;overflow:hidden}.basemap-container .basemap{border:2px solid white;background-color:#fff;border-radius:6px;cursor:pointer}.basemap-container .basemap .image{height:75px;width:75px;position:relative}.basemap-container .basemap .image .name{font-size:10px;line-height:15px;text-align:center;position:absolute;width:100%;bottom:0;background-color:#f5f5f580;border-radius:0 0 6px 6px;text-shadow:1px 1px 2px white;text-overflow:ellipsis;white-space:nowrap;overflow:hidden}.basemap-container .basemap .image img{width:75px;border-radius:6px;image-rendering:auto}.basemap-container .basemap .image .no-image{width:75px;height:75px;background-color:#fff;text-align:center;padding-top:20px}.basemap-container .selected{border:2px solid #337ab7;background-color:#337ab7;color:#337ab7;font-weight:600}\n"] }]
        }], ctorParameters: () => [{ type: i1.BasemapService }, { type: i2.AbstractArlasMapService }, { type: i3.ArlasMapFrameworkService }], propDecorators: { map: [{
                type: Input
            }], mapSources: [{
                type: Input
            }], basemapChanged: [{
                type: Output
            }], blur: [{
                type: Output
            }] } });
