// arlas-web-components/esm2022/lib/services/collection.service.mjs
/*
 * Licensed to Gisaïa under one or more contributor
 * license agreements. See the NOTICE.txt file distributed with
 * this work for additional information regarding copyright
 * ownership. Gisaïa licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class BaseCollectionService {
}
export class AwcCollectionService extends BaseCollectionService {
    getAllUnits() {
        return [];
    }
    _initUnits() { }
    _initDisplayNames() { }
    getDisplayName(collectionName) {
        return collectionName;
    }
    getDisplayFieldName(fieldName) {
        return fieldName;
    }
    getUnit(collectionName) {
        return collectionName;
    }
}
export class CollectionService {
    constructor(baseCollectionService) {
        this.baseCollectionService = baseCollectionService;
    }
    getDisplayName(collectionName) {
        return this.baseCollectionService.getDisplayName(collectionName);
    }
    ;
    getDisplayFieldName(fieldName) {
        return this.baseCollectionService.getDisplayFieldName(fieldName);
    }
    ;
    getUnit(collectionName) {
        return this.baseCollectionService.getUnit(collectionName);
    }
    ;
    getAllUnits() {
        return this.baseCollectionService.getAllUnits();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: CollectionService, deps: [{ token: BaseCollectionService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: CollectionService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: CollectionService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: () => [{ type: BaseCollectionService }] });
