// arlas-web-components/esm2022/lib/components/powerbars/powerbar/powerbar.module.mjs
/*
 * Licensed to Gisaïa under one or more contributor
 * license agreements. See the NOTICE.txt file distributed with
 * this work for additional information regarding copyright
 * ownership. Gisaïa licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { NgModule } from '@angular/core';
import { PowerbarComponent } from './powerbar.component';
import { MatTooltipModule } from '@angular/material/tooltip';
import { TranslateModule } from '@ngx-translate/core';
import { FormatNumberModule } from '../../../pipes/format-number/format-number.module';
import { ShortenNumberModule } from '../../../pipes/shorten-number/shorten-number.module';
import { CommonModule } from '@angular/common';
import { MatCheckboxModule } from '@angular/material/checkbox';
import * as i0 from "@angular/core";
export class PowerbarModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: PowerbarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.13", ngImport: i0, type: PowerbarModule, declarations: [PowerbarComponent], imports: [CommonModule,
            MatTooltipModule,
            MatCheckboxModule,
            TranslateModule,
            ShortenNumberModule,
            FormatNumberModule], exports: [PowerbarComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: PowerbarModule, imports: [CommonModule,
            MatTooltipModule,
            MatCheckboxModule,
            TranslateModule,
            ShortenNumberModule,
            FormatNumberModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: PowerbarModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        MatTooltipModule,
                        MatCheckboxModule,
                        TranslateModule,
                        ShortenNumberModule,
                        FormatNumberModule
                    ],
                    declarations: [PowerbarComponent],
                    exports: [PowerbarComponent]
                }]
        }] });
