// arlas-web-components/esm2022/lib/components/powerbars/powerbars.module.mjs
/*
 * Licensed to Gisaïa under one or more contributor
 * license agreements. See the NOTICE.txt file distributed with
 * this work for additional information regarding copyright
 * ownership. Gisaïa licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MatCardModule } from '@angular/material/card';
import { MatIconModule } from '@angular/material/icon';
import { MatInputModule } from '@angular/material/input';
import { MatTooltipModule } from '@angular/material/tooltip';
import { TranslateModule } from '@ngx-translate/core';
import { ColorGeneratorModule } from '../../services/color.generator.module';
import { FormatNumberModule } from '../../pipes/format-number/format-number.module';
import { PowerbarsComponent } from './powerbars.component';
import { MatMenuModule } from '@angular/material/menu';
import { MatButtonModule } from '@angular/material/button';
import { MatButtonToggleModule } from '@angular/material/button-toggle';
import { PowerbarModule } from './powerbar/powerbar.module';
import * as i0 from "@angular/core";
import * as i1 from "../../services/color.generator.module";
export class PowerbarsModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: PowerbarsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.2.13", ngImport: i0, type: PowerbarsModule, declarations: [PowerbarsComponent], imports: [CommonModule,
            MatCardModule,
            MatIconModule,
            MatTooltipModule,
            MatMenuModule,
            MatInputModule,
            MatButtonModule,
            MatButtonToggleModule, i1.ColorGeneratorModule, TranslateModule,
            PowerbarModule,
            FormatNumberModule], exports: [PowerbarsComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: PowerbarsModule, imports: [CommonModule,
            MatCardModule,
            MatIconModule,
            MatTooltipModule,
            MatMenuModule,
            MatInputModule,
            MatButtonModule,
            MatButtonToggleModule,
            ColorGeneratorModule.forRoot(),
            TranslateModule,
            PowerbarModule,
            FormatNumberModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: PowerbarsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule,
                        MatCardModule,
                        MatIconModule,
                        MatTooltipModule,
                        MatMenuModule,
                        MatInputModule,
                        MatButtonModule,
                        MatButtonToggleModule,
                        ColorGeneratorModule.forRoot(),
                        TranslateModule,
                        PowerbarModule,
                        FormatNumberModule
                    ],
                    declarations: [PowerbarsComponent],
                    exports: [PowerbarsComponent]
                }]
        }] });
