// arlas-web-components/esm2022/lib/components/powerbars/model/powerbar.mjs
/*
 * Licensed to Gisaïa under one or more contributor
 * license agreements. See the NOTICE.txt file distributed with
 * this work for additional information regarding copyright
 * ownership. Gisaïa licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
export class PowerBar {
    constructor(term, parentTerm, count) {
        /**
         * @description Whether the powerbar is selected.
         */
        this.isSelected = false;
        /**
         * @description class name to apply to the powerbar : `neutral-state`, `selected-bar` or `unselected-bar`.
         */
        this.classSuffix = 'neutral-state';
        this.term = term;
        this.parentTerm = parentTerm;
        this.count = count;
    }
}
