// arlas-web-components/esm2022/lib/pipes/get-collection-display-name/get-collection-display-name.pipe.mjs
/*
 * Licensed to Gisaïa under one or more contributor
 * license agreements. See the NOTICE.txt file distributed with
 * this work for additional information regarding copyright
 * ownership. Gisaïa licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { Pipe } from '@angular/core';
import { CollectionService } from '../../services/collection.service';
import * as i0 from "@angular/core";
import * as i1 from "../../services/collection.service";
export class GetCollectionDisplayNamePipe {
    constructor(collectionService) {
        this.collectionService = collectionService;
    }
    transform(value, ...args) {
        if (!value) {
            return '';
        }
        return this.collectionService.getDisplayName(value);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: GetCollectionDisplayNamePipe, deps: [{ token: i1.CollectionService }], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.2.13", ngImport: i0, type: GetCollectionDisplayNamePipe, name: "getCollectionDisplayName" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.13", ngImport: i0, type: GetCollectionDisplayNamePipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'getCollectionDisplayName'
                }]
        }], ctorParameters: () => [{ type: i1.CollectionService }] });
