// arlas-map/esm2022/lib/legend/legend.config.mjs
/*
 * Licensed to Gisaïa under one or more contributor
 * license agreements. See the NOTICE.txt file distributed with
 * this work for additional information regarding copyright
 * ownership. Gisaïa licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
export var PROPERTY_SELECTOR_SOURCE;
(function (PROPERTY_SELECTOR_SOURCE) {
    PROPERTY_SELECTOR_SOURCE["fix"] = "Fix";
    PROPERTY_SELECTOR_SOURCE["provided"] = "Provided";
    PROPERTY_SELECTOR_SOURCE["generated"] = "Generated";
    PROPERTY_SELECTOR_SOURCE["manual"] = "Manual";
    PROPERTY_SELECTOR_SOURCE["interpolated"] = "Interpolated";
    PROPERTY_SELECTOR_SOURCE["metric_on_field"] = "Metric on field";
    PROPERTY_SELECTOR_SOURCE["heatmap_density"] = "Density";
})(PROPERTY_SELECTOR_SOURCE || (PROPERTY_SELECTOR_SOURCE = {}));
