// arlas-web-components/esm2022/lib/components/results/utils/results.utils.mjs
/*
 * Licensed to Gisaïa under one or more contributor
 * license agreements. See the NOTICE.txt file distributed with
 * this work for additional information regarding copyright
 * ownership. Gisaïa licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
/**
 * Enables customize the resultlist options/behaviours
 */
export class ResultListOptions {
    constructor() {
        this.showActionsOnhover = false;
        this.showDetailIconName = 'add_circle_outline';
        this.hideDetailIconName = 'remove_circle_outline';
        this.defautlImgUrl = './assets/no-view.png';
    }
}
/** A utility class to handle the action's states  */
export class ActionHandler {
    /** An action is reversible simply when a reverse action is declared. */
    static isReversible(a) {
        return !!a && !!a.reverseAction;
    }
    static activate(a) {
        a.activated = true;
    }
    static reverse(a) {
        a.activated = false;
    }
}
export const PROTECTED_IMAGE_HEADER = 'Protected-Image-Header';
/**
 * @param data A dictionnary of data to retrieve information
 * @param template The template of the desired string. Contains variable keys between brackets
 * @returns A string with the regex replaced by the data
 */
export function matchAndReplace(data, template) {
    let replaced = template;
    template.match(/{(.+?)}/g)?.forEach(t => {
        const key = t.replace('{', '').replace('}', '');
        if (!!data.get(key)) {
            replaced = replaced.replace(t, data.get(key).toString());
        }
    });
    return replaced;
}
